### PR TITLE
Upgrade Errors for struct enum functions

### PR DESCRIPTION
**Pull request by jordanjennings-mysten**
## Description 

Add struct, enum and function errors to upgrade errors, leaves out type params to give it, it's own PR.

## Test plan 

snapshots

---

## Release notes

no effect until enabled

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui/src/lib.rs [hunk 1]
```diff
@@ -15,6 +15,7 @@ pub mod key_identity;
 pub mod keytool;
 pub mod shell;
 pub mod sui_commands;
+mod upgrade_compatibility;
```
Comment (tnowacki) [inline crates/sui/src/lib.rs:18]: What does this change do?

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+31 │     public enum EnumChangeVariant {
+   │                 ----------------- Enum definition
+32 │         A,
+33 │         C, // changed from B
+   │         ^ Mismatched variant name 'C', expected 'B'.
+   │
+   = Enums are part of a module's public interface and cannot be changed during an upgrade
+   = Restore the original enum's variants for enum 'EnumChangeVariant' including the ordering.
+
+error[Compatibility E01002]: type mismatch
```
Comment (tnowacki) [inline crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap:65]: Is this really a "type mismatch"?

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+   ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/enum_errors_v2/sources/UpgradeErrors.move:59:9
+   │
+57 │     public enum EnumAddVariantWithTypes {
+   │                 ----------------------- Enum definition
+58 │         A { a: u8 },
+59 │         B { b: u8 }, // added
+   │         ^^^^^^^^^^^ New unexpected variant 'B'.
+   │
+   = Enums are part of a module's public interface and cannot be changed during an upgrade.
```
Comment (tnowacki) [inline crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap:149]: I'm seeing some weird characters at the end of these lines on GitHub. Any idea what is up here?

File: crates/sui/src/upgrade_compatibility.rs [hunk 4]
```diff
@@ -513,85 +572,722 @@ fn compare_packages(
...
+    // handle function arguments
+    if old_function.parameters.len() != new_function.parameters.len() {
+        diags.add(
+            Diagnostic::new(
+                Function_::SignatureMismatch,
+                (def_loc, format!("Expected {} parameters, have {}", old_function.parameters.len(), new_function.parameters.len())),
+                Vec::<(Loc, String)>::new(),
+                vec![
+                    "Functions are part of a module's public interface and cannot be changed during an upgrade.".to_string(),
+                    format!("Restore the original function's parameters for function '{function_name}', expected {} parameters.", old_function.parameters.len())
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:752]: You might want to do some manual line breaks in the string here for some manual formatting, e.g.
```suggestion
                    format!("Restore the original function's parameters for \
                        function '{function_name}', expected {} parameters.", old_function.parameters.len())
```

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap (hunk 3 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap:149]: the utf8 character code for the period is `0x2e` which is a regular period 🤷 guessing highlighting feature in github is having an issue, single quotes similarly are also throwing it off.
`cat -e` seems to suggest the newlines are fine.
image

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap (hunk 2 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__enum.snap:65]: The only other idea I had was layout mismatch?

**Review by tnowacki — APPROVED** [review]

**Pull request merged**